res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,741 @@
+};
+
+PioneerDDJ400.beatjumpShiftDown = function(_channel, control, value, _status, group) {
+    if (value === 0 || PioneerDDJ400.beatjumpPad[0x21] / 16 < 1/16) {
+        return;
+    }
+    Object.keys(PioneerDDJ400.beatjumpPad).forEach(function(pad) {
+        PioneerDDJ400.beatjumpPad[pad] = PioneerDDJ400.beatjumpPad[pad] / 16;
+    });
+    engine.setValue(group, "beatjump_size", PioneerDDJ400.beatjumpPad[0x21]);
+};
+
+//
+// Sampler mode
+//
+
+PioneerDDJ400.samplerPlayOutputCallbackFunction = function(value, group, _control) {
+    if (value === 1) {
+        var curPad = group.match(/^\[Sampler(\d+)\]$/)[1];
+        PioneerDDJ400.startSamplerBlink((0x97 + (curPad > 8 ? 2 : 0)), (0x30 + ((curPad > 8 ? curPad-8 : curPad)-1)), group);
+    }
+};
+
+PioneerDDJ400.samplerModePadPressed = function(_channel, _control, value, _status, group) {
+    if (engine.getValue(group, "track_loaded")) {
+        engine.setValue(group, "cue_gotoandplay", value);
+    } else {
+        engine.setValue(group, "LoadSelectedTrack", value);
+    }
+};
+
+PioneerDDJ400.samplerModeShiftPadPressed = function(_channel, _control, value, _status, group) {
+    if (engine.getValue(group, "play")) {
+        engine.setValue(group, "cue_gotoandstop", value);
+    } else if (engine.getValue(group, "track_loaded")) {
+        engine.setValue(group, "eject", value);
+    }
+};
+
+PioneerDDJ400.startSamplerBlink = function(channel, control, group) {
+    var val = 0x7f;
+
+    PioneerDDJ400.stopSamplerBlink(channel, control);
+    PioneerDDJ400.timers[channel][control] = engine.beginTimer(250, function() {
+        val = 0x7f - val;
+
+        // blink the appropriate pad
+        midi.sendShortMsg(channel, control, val);
+        // also blink the pad while SHIFT is pressed
+        midi.sendShortMsg((channel+1), control, val);
+
+        var isPlaying = engine.getValue(group, "play") === 1;
+
+        if (!isPlaying) {
+            // kill timer
+            PioneerDDJ400.stopSamplerBlink(channel, control);
+            // set the pad LED to ON
+            midi.sendShortMsg(channel, control, 0x7f);
+            // set the pad LED to ON while SHIFT is pressed
+            midi.sendShortMsg((channel+1), control, 0x7f);
+        }
+    });
+};
+
+PioneerDDJ400.stopSamplerBlink = function(channel, control) {
+    PioneerDDJ400.timers[channel] = PioneerDDJ400.timers[channel] || {};
+
+    if (PioneerDDJ400.timers[channel][control] !== undefined) {
+        engine.stopTimer(PioneerDDJ400.timers[channel][control]);
+        PioneerDDJ400.timers[channel][control] = undefined;
+    }
+};
+
+//
+// Additional features
+//
+
+PioneerDDJ400.toggleQuantize = function(_channel, _control, value, _status, group) {
+    if (value) {
+        script.toggleControl(group, "quantize");
+    }
+};
+
+PioneerDDJ400.phraseJumpForward = function(_channel, _control, value, _status, group) {
+    if (value) {
+        engine.setValue(group, "beatjump", 8 * 4);
```

Comment:
```suggestion
        engine.setValue(group, "beatjump", 32);
```

Reply:
I'd prefer the 8 * 4 notation, as 32 is more of a 'Magic number'. Same for the other place(s)

Reply:
`8 * 4` is no more or less a magic number than `32`.

Reply:
Come to think of it, should this use `beatjump_size` rather than an arbitrary fixed number?

Reply:
I'm not sure that's a great idea without a way to change the beatjump size from the controller without jumping.

Reply:
Storing it in a descriptive variable could never hurt, especially when it occurs in several places, so I wholeheartly agree!

Reply:
> I'm not sure that's a great idea without a way to change the beatjump size from the controller without jumping.

One can use Beatjump mode to change sizes and jump around differently. This little extra mapping is however, far more convenient and useful for track prepping.

> Come to think of it, should this use beatjump_size rather than an arbitrary fixed number?

Good point. Will add a variable.